aboleth/layers.py

```diff
@@ -354,6 +354,136 @@ def _transformation(self, XP):
+        This is the prior distribution object to use on the layer weights. It
+        must have parameters compatible with (input_dim, output_dim) shaped
+        weights. This ignores the ``std`` parameter.
+    prior_b : tf.distributions.Distribution, optional
+        This is the prior distribution object to use on the layer intercept. It
+        must have parameters compatible with (output_dim,) shaped weights.
+        This ignores the ``std`` and ``use_bias`` parameters.
+    post_W : tf.distributions.Distribution, optional
+        It must have parameters compatible with (input_dim, output_dim) shaped
+        weights. This ignores the ``full`` parameter. See also
+        ``distributions.gaus_posterior``.
+    post_b : tf.distributions.Distributions, optional
+        This is the posterior distribution object to use on the layer
+        intercept. It must have parameters compatible with (output_dim,) shaped
+        weights. This ignores the ``use_bias`` parameters.  See also
+        ``distributions.norm_posterior``.
+
+    """
+
+    def __init__(self, filters, kernel_size, strides=(1, 1), padding='SAME',
+                 std=1., use_bias=True, prior_W=None, prior_b=None,
+                 post_W=None, post_b=None):
+        """Create and instance of a variational Conv2D layer."""
+        self.filters = filters
+        self.kernel_size = kernel_size
+        self.strides = [1] + list(strides) + [1]
+        self.padding = padding
+        self.std = std
+        self.use_bias = use_bias
+        self.pW = prior_W
+        self.pb = prior_b
+        self.qW = post_W
+        self.qb = post_b
+
+    def _build(self, X):
+        """Build the graph of this layer."""
+        n_samples, (height, width, channels) = self._get_X_dims(X)
+        W_shape, b_shape = self._weight_shapes(channels)
+
+        # Layer weights
+        self.pW = self._make_prior(self.pW, W_shape)
+        self.qW = self._make_posterior(self.qW, W_shape)
+
+        # Regularizers
+        KL = kl_sum(self.qW, self.pW)
+
+        # Linear layer
+        Wsamples = self._sample_W(self.qW, n_samples)
+        Net = tf.map_fn(
+            lambda args: tf.nn.conv2d(*args,
+                                      padding=self.padding,
+                                      strides=self.strides),
+            elems=(X, Wsamples), dtype=tf.float32)
+
+        # Optional bias
+        if self.use_bias or not (self.prior_b is None and self.post_b is None):
+            # Layer intercepts
+            self.pb = self._make_prior(self.pb, b_shape)
+            self.qb = self._make_posterior(self.qb, b_shape)
+
+            # Regularizers
+            KL += kl_sum(self.qb, self.pb)
+
+            # Linear layer
+            bsamples = tf.reshape(self._sample_W(self.qb, n_samples),
+                                  [n_samples, 1, 1, 1, self.filters])
+            Net += bsamples
+
+        return Net, KL
+
+    def _make_prior(self, prior_W, weight_shape):
```

Comment:
Should we make this a module utility function since the DenseVariational layer uses this as well? Or something else we can do to make sure DRY?

Reply:
Yeah I think it would be a good idea parameterized layers, just need to modify the function argument to specify whether to use full cov Gaussian since not all layers will be able to support this so they won't have a `self.full` attribute.

Reply:
Cool, I'll make an issue to this affect -- there are probably more places in this module where we could do this too